### ISSUE TITLE
Ability to preserve ordering in key-value collections

### ISSUE THREAD

**Issue by ilg** [cherrywoods/swift-meta-serialization#12]
Is there a mechanism to preserve the order of key-value collections?  I'm thinking about how JSON and other serialization formats often have an inherent order to the key-value pairs in an object/dictionary, but Swift's `Dictionary` doesn't preserve that order, so I'm wondering if there's a way to use `DictionaryLiteral` instead, since it does preserve order.

It would also be nice if it were possible for encodable types to get encoded in a way that preserved the order of their properties, perhaps converting to `DictionaryLiteral` instead of `Dictionary` as the Meta step.  I think Apple's `JSONEncoder` preserves property order by default, which is what makes me think this might be possible; it definitely provides an option to sort the properties by name.

**Comment by cherrywoods** [comment]
Sure, look at how Dictionary is extended to conform to MetaKeyedDecodingContainer and MetaKeyedEncodingContainer. You will need to extend DictionaryLiteral the same way. I have no experience with DictionaryLiteral, but maybe trying an Array of Key values pairs might also be worth looking at... Thanks for your question. The last time I looked at JSONEncoder it used NSDictionary for keyed containers. I don't think NSDictionary preserves order, so I don't think JSONEncoder preserves order, but I haven't looked at this file for some time. At least it is possible to do it.

**Comment by cherrywoods** [comment]
Always fell free to ask more questions!

**Comment by ilg** [comment]
Just to put the information out there in case anyone else comes looking, I think I've got this working using `OrderedDictionary` from `OrderedCollections` in the [swift-collections](https://github.com/apple/swift-collections) package.  I'm using `SimpleSerialization` and changed from a plain `PrimitivesProtocolTranslator` to:
```swift
struct MetaTranslator: MetaSupplier & Unwrapper {
    private let underlyingTranslator = PrimitivesProtocolTranslator<Primitive, Primitive.Type>()
    
    func wrap<T>(_ value: T, for encoder: MetaEncoder) throws -> Meta? where T : Encodable {
        return try underlyingTranslator.wrap(value, for: encoder)
    }
    
    func unwrap<T>(meta: Meta, toType type: T.Type, for decoder: MetaDecoder) throws -> T? where T : Decodable {
        return try underlyingTranslator.unwrap(meta: meta, toType: type, for: decoder)
    }
    
    func keyedContainerMeta() -> EncodingKeyedContainerMeta {
        return OrderedDictionary<String, Meta>()
    }
}
```
and extended `OrderedDictionary` to conform to `KeyedContainerMeta` and `Meta`, basically identically to the `Dictionary` extension in `MetaSerialization`:
```swift
extension OrderedDictionary: KeyedContainerMeta, Meta where Key == String, Value == Meta {
    public func contains(key: MetaCodingKey) -> Bool {
        return self[key.stringValue] != nil
    }
    
    public func getValue(for key: MetaCodingKey) -> Meta? {
        return self[key.stringValue]
    }
    
    public var allKeys: [MetaCodingKey] {
        return keys.map { MetaCodingKey(stringValue: $0) }
    }
    
    public mutating func put(_ value: Meta, for key: MetaCodingKey) {
        self[key.stringValue] = value
    }
}
```

So far, this seems to cause structs to encode with their properties consistently remaining in the order in which they were declared in the struct.

(I hadn't worked on the project in which I was using MetaSerialization in some time.)

**Comment by cherrywoods** [comment]
Thanks! It's nice to hear that someone is still using the library. Your implementation looks good. 
Actually, I think it would make sense to use `OrderedDictionary` as the default keyed container type. Would you be interested in submitting your code as a PR?

**Comment by ilg** [comment]
I'd definitely be interested in contributing this change, but I think there are a couple issues.
1. It'd add a dependency on Apple's swift-collections package.  I don't know if it'd be better to try to avoid that kind of external dependency.
2. If I do add that dependency, I'm rusty enough with cocoapods and carthage that I'm not immediately sure of the best way to go about adding the dependency and testing it.

**Comment by ilg** [comment]
[Here's a quick first pass](https://github.com/ilg/swift-meta-serialization/compare/fix-spm...ilg:swift-meta-serialization:OrderedDictionary?expand=1) at using `OrderedDictionary` (though only through SPM, so probably breaking the Xcode project and CocoaPods and Carthage).

**Comment by cherrywoods** [comment]
> I'd definitely be interested in contributing this change, but I think there are a couple issues.
> 
>     1. It'd add a dependency on Apple's swift-collections package.  I don't know if it'd be better to try to avoid that kind of external dependency.
> 
>     2. If I do add that dependency, I'm rusty enough with cocoapods and carthage that I'm not immediately sure of the best way to go about adding the dependency and testing it.

I think adding the dependency is fine. It's stable and it seems standard enough to me to me (but I haven't developed in Swift for a while). 
Thanks for contributing this, I'll hopefully find time soon to look into integrating this and looking into Carthage and CocoaPods.

**Comment by cherrywoods** [comment]
Reopening this to make it easier to find for me.

**Comment by cherrywoods** [comment]
I changed my mind a bit abound the additional dependency: I added a small new keyed container class that maintains the order keys are inserted: https://github.com/cherrywoods/swift-meta-serialization/blob/main/Sources/MetaSerialization/Frontends%20and%20Default%20Implementations/OrderedKeyedContainer.swift 
This way the repository has fewer dependencies but still offers maintaining the insertion order. `PrimitivesEnumTranslator` also offers using this container for now using the `maintainOrder` argument of it's initializer.

@ilg your solution still works perfectly fine and I don't think there would be any particular reason to prefer `OrderedKeyedContainer` over your solution.
